### PR TITLE
Fixes for missing avatar on record replays

### PR DESCRIPTION
**Pull request by SimonWalton-HiFi**
https://highfidelity.manuscript.com/f/cases/12103/

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/7733/app_12564_PR.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/7733/HighFidelity-Beta-PR12564-8fb3015ec8677b71f8315f62ec83e0501c18d6ff.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/7733/HighFidelity-Beta-PR12564-8fb3015ec8677b71f8315f62ec83e0501c18d6ff.exe

File: assignment-client/src/avatars/AvatarMixer.cpp [hunk 1]
```diff
@@ -605,7 +605,9 @@ void AvatarMixer::handleAvatarIdentityPacket(QSharedPointer<ReceivedMessage> mes
 
 void AvatarMixer::handleKillAvatarPacket(QSharedPointer<ReceivedMessage> message, SharedNodePointer node) {
     auto start = usecTimestampNow();
-    DependencyManager::get<NodeList>()->processKillNode(*message);
+    node->stopPingTimer();
+    emit(DependencyManager::get<NodeList>()->nodeKilled(node));
```
Comment (birarda) [inline assignment-client/src/avatars/AvatarMixer.cpp]: Do we want this behaviour as well if we're simply clearing the node linked data now?

File: libraries/networking/src/udt/Socket.cpp [hunk 2]
```diff
@@ -402,6 +402,10 @@ void Socket::readPendingDatagrams() {
                                                                                   packet->getDataSize(),
                                                                                   packet->getPayloadSize())) {
                         // the connection could not be created or indicated that we should not continue processing this packet
+#ifdef UDT_CONNECTION_DEBUG
+                        qCDebug(networking) << "Bad packet version" << (unsigned int)NLPacket::versionInHeader(*packet)
+                            << ", type" << NLPacket::typeInHeader(*packet);
```
Comment (birarda) [inline libraries/networking/src/udt/Socket.cpp]: This debug makes we think it's a bad packet version, not showing the version of a bad packet. Can we make it a bit clearer - ex: "Received a packet that could not be processed - " and then the version and type after.

File: assignment-client/src/avatars/AvatarMixer.cpp (hunk 1 above)
Comment (SimonWalton-HiFi) [reply assignment-client/src/avatars/AvatarMixer.cpp]: Without the signal the avatar does not immediately disappear but freezes on KillAvatar.

File: libraries/networking/src/udt/Socket.cpp (hunk 2 above)
Comment (SimonWalton-HiFi) [reply libraries/networking/src/udt/Socket.cpp]: I'll clean that up. In this case processReceivedSequenceNumber() gives a debug message first but not in all cases, eg. duplicates.

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/7737/app_12564_PR.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/7737/HighFidelity-Beta-PR12564-547d86a2d52699d27b0cb553e0e47ecd02a65f2d.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/7737/HighFidelity-Beta-PR12564-547d86a2d52699d27b0cb553e0e47ecd02a65f2d.exe

**Comment by ctrlaltdavid** [comment]
Nice! I tried this PR with the recently updated Record script on the Marketplace and it works: successfully made and played back several recordings on a single AC playback script, with all play backs having both sound and avatar.

File: assignment-client/src/avatars/AvatarMixer.cpp (hunk 1 above)
Comment (birarda) [reply assignment-client/src/avatars/AvatarMixer.cpp]: I see - there's logic inside `AvatarMixer::nodeKilled` we need that gets sent down to other nodes.

Can you pull that out into a `handleAvatarKilled` that takes a `Node&` and then call that from here before calling `node->setLinkedData(nullptr)` (and from `AvatarMixer::nodeKilled` as well).

We should also not need `stopPingTimer` right?

Comment (SimonWalton-HiFi) [reply assignment-client/src/avatars/AvatarMixer.cpp]: I've put that in & taken out ```stopPingTimer``` and it looks good. Is there anything ```AvatarMixer::nodeKilled``` should do apart from chain to ```AvatarMixer::handleAvatarKilled```?

Comment (birarda) [reply assignment-client/src/avatars/AvatarMixer.cpp]: Nope, in fact you could probably just change the connection from NodeList::nodeKilled to handleAvatarKilled (which I guess would have been the equivalent of renaming `nodeKilled` in `AvatarMixer` for clarity. I guess it's a bit weird to do the `killedNode->getType() == NodeType::Agent` check when you should already have an avatar, but that shouldn't be an issue.

**Comment by birarda** [comment]
#### Test Plan

@ctrlaltdavid can you re-test this PR, confirming that it works with the expected record app and when the recording restarts the avatar immediately disappears when it is stopped and re-appears when restarted?

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/7759/app_12564_PR.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/7759/HighFidelity-Beta-PR12564-30206ff33e2a4275cc0f4e82029f52c4d341929b.dmg

**Comment by SimonWalton-HiFi** [comment]
A minimal hotfix for RC64 has been created:
https://github.com/highfidelity/hifi/pull/12576

**Comment by SimonWalton-HiFi** [comment]
Test plan:
Repro steps from Manuscript:

Using record.js app and associated playRecordingAc.js in \unpublishedScripts\marketplace\record.

>Simple repro:
>- Make a recording with avatar movement and sound.
>- Run 1 copy of playRecordingAC.js on the server.
>- Load the recording. ... Avatar movement and sound plays back.
>- Stop the recording.
>- Load the recording.

Observe debug output from Agent AC and Avatar-mixer AC.

Before fix:
The recorded avatar will not display upon replay.
When the first replay is stopped the Avatar-mixer log will show:
> Killed "Agent" ...
> ...
> Added "Agent" ...

After fix:
The recorded avatar will replay as many times as requested.
The killed & added Agent messages will not appear in the avatar-mixer log.

Node kick behaviour: TBD

**Comment by birarda** [comment]
build this please

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/7990/app_12564_PR.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/7990/HighFidelity-Beta-PR12564-30206ff33e2a4275cc0f4e82029f52c4d341929b.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/7990/HighFidelity-Beta-PR12564-30206ff33e2a4275cc0f4e82029f52c4d341929b.exe

**Comment by conklin94122** [comment]
Removing QA Ready as changes have been requested. Once those changes are make please reapply QA Ready.

**Comment by birarda** [comment]
@conklin94122 changes were made and I just looked and approved, re-marking QA Ready

**Comment by qlimnavarro** [comment]
Currently Testing.

**Comment by qlimnavarro** [comment]
Pass - @conklin94122 
Logs for domain look as expected.

**Pull request merged**